### ISSUE TITLE
discuss submodules versus subrepo

### ISSUE THREAD

**Issue by jcolomb** [tonic-team/tonic.site#10]
To @tamarachamind, git subrepo seems to be more suitable for modularization in this kind of projects than submodules: https://github.com/ingydotnet/git-subrepo. I can elaborate my reasons a bit further.

**Comment by tamaracha** [comment]
Yes, I'm interested in your thoughts and findings about this topic.

**Comment by tamaracha** [comment]
Sorry, have seen right now that you quoted my message on the slack channel

**Comment by tamaracha** [comment]
My view about modularization and submodules:

Anyway, what we want is to enable data re-use across research projects.
I can identify two modularization perspectives one can take in this matter:

1. Somebody wants to depend on another project (passive perspective).
2. Somebody wants to share a subset of their own project (active perspective).

The passive scenario is relatively straight forward, just link your dependency in and you're all set.
That's where submodules can be an acceptable solution for dependency management.
These are just links to a specific commit or branch of a given repo, the link will be broken if the submodule repo has moved.
They add some deviations from the normal git workflows (`git clone --recurse-submodules`), but it can be manageable.

For the active scenario, I can't recommend submodules.
They feel like ordinary directories, but they are something different which puts some cognitive load on the user.

- One quickly overlooks that one is currently working in a submodule, then one also has to commit in the parent repo to get it clean again.
- Submodules are mostly in detached head state, so you'll have to remember to checkout a branch in the submodule before committing, or you have to merge your commit afterwards. If commits are not reachable from a branch, they can't be pushed.
- Submodules often break in more elaborate git workflows like rebase.
- Turning a subdirectory into a submodule is complicated.

Personally, I would use [subrepo] for the active scenario in my projects.
It deals better with the issues mentioned above, and you're not dependent of the child repos being available, because it stores a copy of the child repo instead of a link.

[subrepo]: https://github.com/ingydotnet/git-subrepo

**Comment by tamaracha** [comment]
My alternative suggestion would be to have a top-level dependencies folder where dependency projects can be linked as submodules.
Putting deps in a dedicated location makes the relationship clear and keeps users from inadvertently making changes to them.
In this approach, splitting out project parts as submodule/subrepo wouldn't be necessary,.
This makes a project more maintenable (fiewer repos), but you have only access rights management for the full project.

**Comment by jcolomb** [comment]
hi @tamaracha, thanks for sharing.

I am a bit lost in the computer science vocabulary. The use case I am thinking of is somewhat different from the active and passive scenariis you mentionned, and I will need to elaborate that (see #11) so we can indeed discuss this point further.

**Comment by jcolomb** [comment]
Another point is that Datalad is using submodules (and not subrepo if I got it right) so we may want to follow that path to be compatible with Datalad.
